tests/test_flows.py

```diff
@@ -0,0 +1,356 @@
+        self.message(message, 35)
+
+    def err(self, message: str):
+        '''Print a red error message.'''
+        self.message(message, 31)
+
+    def ok(self):
+        '''Print a green OK message.'''
+        self.message("OK", 32)
+
+    def fail(self, message: str) -> bool:
+        '''Print a red error message, increment failure count and return False.'''
+        self.check_error += 1
+        self.err('FAIL: {}'.format(message))
+        if self.opts.stop_on_failure:
+            sys.exit(1)
+        return False
+
+    def start_test(self, test: str):
+        '''Call this at the beginning of a test.'''
+        self.running = test
+        self.status("starting")
+
+    def tcpdump_start(self):
+        '''Start tcpdump if enabled by config.'''
+        if self.opts.tcpdump:
+            outputfile = os.path.join(self.rootdir, 'logs', 'test_{}.pcap'.format(self.running))
+            self.debug("Starting tcpdump, output file {}".format(outputfile))
+            inspect = json.loads(subprocess.check_output(('docker', 'network', 'inspect',
+                                                          'prplMesh-net-{}'.format(self.opts.unique_id))))
+            bridge = inspect[0]['plugins'][0]['bridge']
+            self.tcpdump_proc = subprocess.Popen(["tcpdump", "-i", bridge, "-w", outputfile], stderr=subprocess.PIPE)
+            # tcpdump takes a while to start up. Wait for the appropriate output before continuing.
+            # poll() so we exit the loop if tcpdump terminates for any reason.
+            while not self.tcpdump_proc.poll():
+                line = self.tcpdump_proc.stderr.readline()
+                self.debug(line.decode()[:-1]) # strip off newline
+                if line.startswith(b"tcpdump: listening on " + bridge.encode()):
+                    self.tcpdump_proc.stderr.close() # Make sure it doesn't block due to stderr buffering
+                    break
+            else:
+                self.err("tcpdump terminated")
+                self.tcpdump_proc = None
+
+    def tcpdump_kill(self):
+        '''Stop tcpdump if it is running.'''
+        if self.tcpdump_proc:
+            self.status("Terminating tcpdump")
+            self.tcpdump_proc.terminate()
+            self.tcpdump_proc = None
+
+    def docker_command(self, device: str, *command: str) -> bytes:
+        '''Execute `command` in docker container `device` and return its output.'''
+        return subprocess.check_output(("docker", "exec", device) + command)
+
+    def open_CAPI_socket(self, device: str, controller: bool = False) -> send_CAPI_command.UCCSocket:
+        '''Open a CAPI socket to the agent (or controller, if set) on "device".'''
+        # First, get the UCC port from the config file
+        if controller:
+            config_file_name = 'beerocks_controller.conf'
+        else:
+            config_file_name = 'beerocks_agent.conf'
+        with open(os.path.join(self.installdir, 'config', config_file_name)) as config_file:
```

Comment:
As you said we'll need to handle it in a different manner later, I'm just commenting so that we don't forget.

Reply:
Eh, I'm afraid I forgot what the "as I said" is...

Reply:
Actually I don't know why I've put my comment here.
I originally wanted to comment on the `cmd_reply` function but it was not something to fix in this PR (the "as you said" was to open the each socket only once in the beginning of the test suite, and to make sure it gets closed at the end).